sympy/solvers/ode/single.py

```diff
@@ -382,6 +381,137 @@ def inverse(self):
+
+    Where the first partials of `P` and `Q` exist and are continuous in a
+    simply connected region.
+
+    A note: SymPy currently has no way to represent inert substitution on an
+    expression, so the hint ``1st_exact_Integral`` will return an integral
+    with `dy`.  This is supposed to represent the function that you are
+    solving for.
+
+    Examples
+    ========
+
+    >>> from sympy import Function, dsolve, cos, sin
+    >>> from sympy.abc import x
+    >>> f = Function('f')
+    >>> dsolve(cos(f(x)) - (x*sin(f(x)) - f(x)**2)*f(x).diff(x),
+    ... f(x), hint='1st_exact')
+    Eq(x*cos(f(x)) + f(x)**3/3, C1)
+
+    References
+    ==========
+
+    - https://en.wikipedia.org/wiki/Exact_differential_equation
+    - M. Tenenbaum & H. Pollard, "Ordinary Differential Equations",
+      Dover 1963, pp. 73
+
+    # indirect doctest
+
+    """
+    hint = "1st_exact"
+    has_integral = True
+    order = [1]
+
+    def _wilds(self, f, x, order):
+        P = Wild('P', exclude=[f(x).diff(x)])
+        Q = Wild('Q', exclude=[f(x).diff(x)])
+        return P, Q
+
+    def _equation(self, fx, x, order):
+        P, Q = self.wilds()
+        return P + Q*fx.diff(x)
+
+    def _verify(self, fx) -> bool:
+        P, Q = self.wilds()
+        x = self.ode_problem.sym
+        y = Dummy('y')
+
+        m, n = self.wilds_match()
+
+        m = m.subs(fx,y)
+        n = n.subs(fx,y)
+        numerator = cancel(m.diff(y) - n.diff(x))
+        if numerator.is_zero:
+            # Is exact
+            return True
+        else:
+            # The following few conditions try to convert a non-exact
+            # differential equation into an exact one.
+            # References:
+            # 1. Differential equations with applications
+            # and historical notes - George E. Simmons
+            # 2. https://math.okstate.edu/people/binegar/2233-S99/2233-l12.pdf
+
+            # If (dP/dy - dQ/dx) / Q = f(x)
+            # then exp(integral(f(x))*equation becomes exact
+            factor = cancel(numerator/n)
+            variables = factor.free_symbols
+            if len(variables) == 1 and x == variables.pop():
+                factor = exp(Integral(factor).doit())
+                m *= factor
+                n *= factor
+                self._wilds_match[P] = m.subs(y, fx)
+                self._wilds_match[Q] = n.subs(y, fx)
+                return True
+            else:
+                # If (dP/dy - dQ/dx) / -P = f(y)
+                # then exp(integral(f(y))*equation becomes exact
+                factor = cancel(-numerator/m)
+                variables = factor.free_symbols
+                if len(variables) == 1 and y == variables.pop():
+                    factor = exp(Integral(factor).doit())
+                    m *= factor
+                    n *= factor
+                    self._wilds_match[P] = m.subs(y, fx)
+                    self._wilds_match[Q] = n.subs(y, fx)
+                    return True
```

Comment:
Most of this code seems to be duplicated above. Is there not a way of simplifying this?

Reply:
I made a new commit which removes the code duplication. It is better now?